StateDB based Health

tommyp1ckles opened the pull request:
# healthv2: add new module health implementation based on statedb.

Similar to existing implementation in hive/cell/<health/structured>.go, this provides system health data in a tree structure.
However, this seeks to reimplement the health provider such that it is no longer coupled to pkg/hive/cell.
Because this will eventually be run just as any other Agent module, we can use StateDB for the implementation.

For more information about the original tree structured health reporter take a look at code documentation [here](https://github.com/cilium/cilium/blob/main/pkg/hive/cell/structured.go#L59-L94). This seeks to provide a similar underlying structure of data while using a much simpler statedb schema.

### Note: The only intended user impact of these changes is that health data will now be available with `cilium-dbg statedb health` and `cilium-dbg statedb dump`. 

This massively reduces the complexity of the health implementation, as well as provides a more convenient data model for storing health update data.

Furthermore, this fixes the awkward distinction between "module" and "subcomponent" reporting by unifying how these health reports are stored in one place.

pkg/healthv2/provider.go implements a new health provider that uses statedb to store a table of health updates by their fully qualified identifier. This is composed of:

[module-id].[componenet-id]

Where the module ID is a fully qualified ID of submodules, ex:

i.  agent.controlplane.bgpv1

Example `cilium statedb health`output:

```bash
Module                                           Component                                                                                              Level   Message                                                        LastOK                 UpdatedAt              Count
agent.controlplane.endpoint-manager              endpoint-gc                                                                                            OK      endpoint-gc                                                    2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.l2-announcer                  job-l2-announcer lease-gc                                                                              OK      Running                                                        2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.auth                          observer-job-auth gc-identity-events                                                                   OK      Primed                                                         2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.auth                          timer-job-auth gc-cleanup                                                                              OK      Primed                                                         2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.auth                          observer-job-auth request-authentication                                                               OK      Primed                                                         2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.node-manager                  nodes-add                                                                                              OK      Node adds successful                                           2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   3
agent.controlplane.node-manager                  nodes-update                                                                                           OK      Node updates successful                                        2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   1
agent.datapath.node-address                      job-node-address-update                                                                                OK      10.244.1.33 (cilium_host), fd00:10:244:1::3c26 (cilium_host)   2024-03-06T23:27:21Z   2024-03-06T23:27:21Z   0
agent.controlplane.bgp-cp                        job-diffstore-events                                                                                   OK      Running                                                        2024-03-06T23:27:22Z   2024-03-06T23:27:22Z   1
agent.controlplane.stale-endpoint-cleanup        job-endpoint-cleanup                                                                                   OK      Running                                                        2024-03-06T23:27:22Z   2024-03-06T23:27:22Z   0
agent.controlplane.envoy-proxy                   timer-job-version-check                                                                                OK      Primed                                                         2024-03-06T23:27:22Z   2024-03-06T23:27:22Z   0
agent.controlplane.service-manager               job-ServiceReconciler                                                                                  OK      2 NodePort frontend addresses                                  2024-03-06T23:27:22Z   2024-03-06T23:27:22Z   0
agent.datapath.l2-responder                      job-l2-responder-reconciler                                                                            OK      Running                                                        2024-03-06T23:27:22Z   2024-03-06T23:27:22Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-1865 (/).policymap-sync                                                                OK      sync-policymap-1865                                            2024-03-06T23:27:23Z   2024-03-06T23:27:23Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-1865 (/).datapath-regenerate                                                           OK      Endpoint regeneration successful                               2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   2
agent.controlplane.endpoint-manager              cilium-endpoint-162 (local-path-storage/local-path-provisioner-7577fdbbfb-pcffg).policymap-sync        OK      sync-policymap-162                                             2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-162 (local-path-storage/local-path-provisioner-7577fdbbfb-pcffg).datapath-regenerate   OK      Endpoint regeneration successful                               2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   1
agent.controlplane.endpoint-manager              cilium-endpoint-2319 (kube-system/coredns-76f75df574-8m4d8).policymap-sync                             OK      sync-policymap-2319                                            2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-2319 (kube-system/coredns-76f75df574-8m4d8).datapath-regenerate                        OK      Endpoint regeneration successful                               2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   1
agent.controlplane.endpoint-manager              cilium-endpoint-605 (kube-system/coredns-76f75df574-ql8kj).policymap-sync                              OK      sync-policymap-605                                             2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-605 (kube-system/coredns-76f75df574-ql8kj).datapath-regenerate                         OK      Endpoint regeneration successful                               2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   1
agent.controlplane.endpoint-manager              cilium-endpoint-3399 (/).policymap-sync                                                                OK      sync-policymap-3399                                            2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   0
agent.controlplane.endpoint-manager              cilium-endpoint-3399 (/).datapath-regenerate                                                           OK      Endpoint regeneration successful                               2024-03-06T23:27:24Z   2024-03-06T23:27:24Z   1
agent.controlplane.node-manager                  background-sync                                                                                        OK      Node validation successful                                     2024-03-06T23:29:57Z   2024-03-06T23:29:57Z   2
agent.controlplane.daemon                        job-sync-hostips                                                                                       OK      Synchronized                                                   2024-03-06T23:30:22Z   2024-03-06T23:30:22Z   3
agent.controlplane.daemon.ep-bpf-prog-watchdog   ep-bpf-prog-watchdog                                                                                   OK      ep-bpf-prog-watchdog                                           2024-03-06T23:30:22Z   2024-03-06T23:30:22Z   6
agent.datapath.sysctl                            job-reconciler-loop                                                                                    OK      OK, 25 objects                                                 2024-03-06T23:30:41Z   2024-03-06T23:30:41Z   62
agent.controlplane.endpoint-manager              cilium-endpoint-2319 (kube-system/coredns-76f75df574-8m4d8).cep-k8s-sync                               OK      sync-to-k8s-ciliumendpoint (2319)                              2024-03-06T23:30:42Z   2024-03-06T23:30:42Z   20
agent.controlplane.endpoint-manager              cilium-endpoint-162 (local-path-storage/local-path-provisioner-7577fdbbfb-pcffg).cep-k8s-sync          OK      sync-to-k8s-ciliumendpoint (162)                               2024-03-06T23:30:42Z   2024-03-06T23:30:42Z   20
agent.controlplane.endpoint-manager              cilium-endpoint-605 (kube-system/coredns-76f75df574-ql8kj).cep-k8s-sync                                OK      sync-to-k8s-ciliumendpoint (605)                               2024-03-06T23:30:42Z   2024-03-06T23:30:42Z   20
agent.datapath.agent-liveness-updater            timer-job-agent-liveness-updater                                                                       OK      OK (107.75µs)                                                  2024-03-06T23:30:43Z   2024-03-06T23:30:43Z   0

```

Similarily the component id is a ID representing a tree of subcomponents of a module.  Together, these form a tree where each path stores information about the module and component being reported on.  As well, Status updates store the original "two-component" identifier.

This schema is meant to be less opinionated about how to view health data, it is simply a set of health report rows indexed by a identifier path.
Because of this, there is no longer any distinction between a "reporter" (i.e. leaf) and "scope" (i.e. parent node). This means that a reporter can have a status and have child reports.

Initially this will be shimmed into the existing health infrastructure in hive/cell. Ultimately we will remove all that code and refactor health reporters using the external github.com/cilium/hive library.

tommyp1ckles committed:
daemon/cmd/cells: add healthv2 provider to agent infra cells.

This will provide the healthv2 provider which manages injecting cell.Module scoped reporters.
Subsequent commits will use a shim as a first step to replacing the reporter in hive/cell.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
pkg/hive: shim healthv2 into existing hive/cell health.

This is a temporary measure to swap out implementations while waiting for external cilium/hive module hooks.
This implements pkg/hive/cell.(HealthReporter) using pkg/healthv2/types.(HealthReporter) thus switching off the old provider/reporting system.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
healthv2: add new module health implementation based on statedb.

Similar to existing implementation in hive/cell/<health/structured>.go, this provides system health data in a tree structure.
However, this seeks to reimplement the health provider such that it is no longer coupled to pkg/hive/cell.
Because this will eventually be run just as any other Agent module, we can use StateDB for the implementation.

This massively reduces the complexity of the health implementation, as well as provides a more convenient data model for storing health update data.

Furthermore, this fixes the awkward distinction between "module" and "subcomponent" reporting by unifying how these health reports are stored in one place.

pkg/healthv2/provider.go implements a new health provider that uses statedb to store a table of health updates by their fully qualified identifier. This is composed of:

[module-id].[componenet-id]

Where the module ID is a fully qualified ID of submodules, ex:

i.  agent.controlplane.bgpv1

Similarily the component id is a ID representing a tree of subcomponents of a module.  Together, these form a tree where each path stores information about the module and component being reported on.  As well, Status updates store the original "two-component" identifier.

This schema is meant to be less opinionated about how to view health data, it is simply a set of health report rows indexed by a identifier path.
Because of this, there is no longer any distinction between a "reporter" (i.e. leaf) and "scope" (i.e. parent node). This means that a reporter can have a status and have child reports.
Subsequent commits will deal with how to correctly display this in 'cilium status' output.

Initially this will be shimmed into the existing health infrastructure in hive/cell. Ultimately we will remove all that code and refactor health reporters using the external github.com/cilium/hive library.

Notable differences between the original health and healthv2:

* Remove 'unknown' state, this was used when requesting a health reporter that was never reported on.
	In practices, this status was hard to understand and difficult to action.
* To make the reporting code more ergonomic, deriving a subreporter is now part of the HealthReporter interface.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

learnitall commented on api/v1/openapi.yaml:
```diff
@@ -1460,7 +1460,9 @@ definitions:
         type: string
-
+      count:
+        description: How many times in a row has this status been reported.
```
Nit: the description here is formatted as a question:

> How many times in a row has this status been reported?

Can we reword this to be a statement?

```suggestion
        description: The number of consecutive status updates.
```

Another option:

```The number of times in a row that the status has been reported.```

learnitall reviewed (changes requested):
For API I just have a small nit, otherwise LGTM.

tommyp1ckles committed:
cilium-dbg: add support for statedb table list cli command.

Running cilium-dbg statedb health will now show health column data.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

joamaki commented on pkg/healthv2/provider.go:
```diff
@@ -0,0 +1,219 @@
+type HealthProvider struct {
+	db           *statedb.DB
+	stopped      atomic.Bool
+	statusTable  statedb.RWTable[types.Status]
```
Let's make `statedb.Table[types.Status]` available to the whole application (also the indexes). This allows tests to check that the status for example.

joamaki commented on pkg/healthv2/provider.go:
```diff
@@ -0,0 +1,219 @@
+			s.Count = newCount
+			defer func() {
+				logrus.Infof("upserted new status: %s", s)
+				tx.Commit()
```
Prefer doing defer of the abort instead of commit. The log message that we upserted is also printed now in case of failed insert.

tommyp1ckles replied:
That makes sense 👍

tommyp1ckles committed:
hive/cell: remove unused code in health/structured.go.

Some code will remain as it is used for testing in other places.
We will remove this once we have switched completely to healthv2.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
cilium-dbg: remove schema and code related to health module endpoint.

The 'cilium status' cli output now uses the statedb remote table endpoint to fetch health data
Thus we can remove the old module health specific code and openapi schema.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

joestringer commented on pkg/healthv2/cell.go:
Looks like these files need a CODEOWNERS update for ownership, presumably to sig-agent?

dylandreimerink commented on pkg/healthv2/provider.go:
```diff
@@ -0,0 +1,219 @@
+	p.PrimaryIndex = statedb.Index[types.Status, types.HealthID]{
+		Name: "identifier",
+		FromObject: func(s types.Status) index.KeySet {
+			return index.NewKeySet([]byte(s.ID.String()))
```
Perhaps nicer/better to use the existing Strigger index helpers https://github.com/cilium/cilium/blob/main/pkg/statedb/index/string.go#L12

joamaki replied:
I think the idea was to have non-terminated key so prefix searches would work. We discussed with @tommyp1ckles about also changing index.String to not terminate and just rely on the iterators to cut off after length mismatches. Though we can also add e.g. index.StringPrefix which is like index.String but no termination and then use that for FromKey, but perhaps less confusing overall if we just drop the key termination.

tommyp1ckles replied:
@dylandreimerink @joamaki yeah, I'd like to play around adding support for prefix querys, including partial key match. For example something like:

agent.control-plane.endpoint-manager.endpoint-* that might match a set of endpoint specific status such as:

 ```
agent.control-plane.endpoint-manager.endpoint-1234.k8s-sync
agent.control-plane.endpoint-manager.endpoint-1234.datapath-reconcile
agent.control-plane.endpoint-manager.endpoint-111.k8s-sync
agent.control-plane.endpoint-manager.endpoint-111.datapath-reconcile
...
```

dylandreimerink commented on pkg/healthv2/provider.go:
```diff
@@ -0,0 +1,219 @@
+		if !ok {
+			break
+		}
+		logrus.Infof("%s (rev=%d)", s.ID.String(), rev)
```
Shouldn't we want to use a FieldLogger provided by hive? Or pkg/logging ?

tommyp1ckles replied:
Yup, good catch, this was supposed to be a temporary measure while developing the code, ill switch that over.

dylandreimerink commented on pkg/healthv2/provider.go:
```diff
@@ -0,0 +1,219 @@
+			if found && old.Level == s.Level && old.Message == s.Message {
+				newCount = old.Count + 1
+			}
+			s.Count = newCount
```
Isn't this easier?
```suggestion
			old, _, found := p.statusTable.First(tx, p.PrimaryIndex.QueryFromObject(s))
			// If a similar status already exists, increment count, otherwise start back
			// at zero.
			if found && old.Level == s.Level && old.Message == s.Message {
				s.Count = old.Count + 1
			}
```

joamaki commented on pkg/healthv2/provider.go:
```diff
@@ -0,0 +1,219 @@
+		FromObject: func(s types.Status) index.KeySet {
+			return index.NewKeySet([]byte(s.ID.String()))
+		},
+		FromKey: func(k types.HealthID) index.Key {
```
Here as well we can do `index.Stringer[types.HealthID]` (and perhaps Go will be able to infer the type in the future...)

derailed commented on cilium-dbg/cmd/status.go:
```diff
@@ -104,8 +108,30 @@ func statusDaemon() {
+				}
+			}
+			if err := <-errChan; err != nil {
+				Fatalf("LowerBound: %s", err)
```
nit: we should probably beef this up to report additional context vs lowerbound failed

tommyp1ckles replied:
This was based on the cilium-dbg/cmd/statedb.go "statedb" dump code.  But I'll add a better message in case this ever fails.

derailed commented on pkg/health/client/modules.go:
```diff
@@ -23,94 +21,109 @@ const (
+func upsertTree(r *node, report *types.Status, stack []string) {
+	if len(stack) == 1 {
+		name := stack[0]
+		//meta := fmt.Sprintf("%s[%s] %s", name, strings.ToUpper(string(report.Level)), report.Message)
```
Delete?

derailed reviewed (commented):
@tommyp1ckles Nice work!

tommyp1ckles committed:
pkg/healthv2: assign hv2 to sig-foundations.

Adds codeowners entry for healthv2.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

bimmlerd commented on CODEOWNERS:
```diff
@@ -458,6 +458,7 @@ Makefile* @cilium/build
 /pkg/flowdebug/ @cilium/proxy
 /pkg/fqdn/ @cilium/fqdn
 /pkg/fswatcher/ @cilium/sig-datapath @cilium/sig-hubble
+/pkg/healthv2/ @cilium/sig-foundations
```
`healthv2` reads weird to me. For modules, it'd be `health/v2` (see [here](https://go.dev/blog/v2-go-modules)) but I'm not sure what the way to go is for packages in tree. Is the goal to eventually rename this `health` again?

tommyp1ckles replied:
health/v2 is definitely cleaner, however there is already a `pkg/health` that has nothing to do with this code so that may be confusing to have that nested inside there.

Perhaps something like healthprovider may be better. pkg/healthv2 may also mislead someone into thinking this is related to pkg/health.

bimmlerd replied:
Oh! Yeah for some reason I definitely assumed this is version2 of the same pkg >.<

bimmlerd commented on cilium-dbg/cmd/status.go:
```diff
@@ -17,6 +18,9 @@ import (
 	"github.com/cilium/cilium/pkg/command"
 	healthPkg "github.com/cilium/cilium/pkg/health/client"
 	"github.com/cilium/cilium/pkg/health/defaults"
+	"github.com/cilium/cilium/pkg/healthv2/types"
```
duplicate import? (`healthTypes` and `types`)

bimmlerd commented on pkg/hive/cell/module.go:
```diff
@@ -12,6 +12,8 @@ import (
 	"github.com/sirupsen/logrus"
 	"go.uber.org/dig"
 
+	"github.com/cilium/cilium/pkg/healthv2/types"
```
duplicate import?

bimmlerd reviewed (approved):
CODEOWNERS is fine by me, couple of nits

tommyp1ckles committed:
operator/cmd: add statedb and healthv2 to operator hive.

operator/pkg/bgpv2 relies on having a health provider.
This adds a healthv2, as well as its dependency statedb to
fix operator hive.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
reconciler: rewrite health check to use healthv2.

This fixes test failures in statedb/reconciler, this
also replaces usingt the to json hack used in the previous
health implementation.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
nodemanager: rewrite health status test for healthv2.

Sets up hive fixture then queries statedb table directly to
check that expected updates occurred

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
l2announcer: fix test panic due to missing healthprovider.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

derailed commented on cilium-dbg/cmd/status.go:
```diff
@@ -104,8 +107,30 @@ func statusDaemon() {
+					iter,
+					func(obj healthTypes.Status, rev statedb.Revision) error {
+						// Remember the latest revision to query from.
+						revision = rev + 1
```
Where is this used? Since we've already queried the db.

derailed commented on pkg/health/client/modules_test.go:
```diff
@@ -5,45 +5,36 @@ package client_test
-│           └── dork                                            [DEGRADED] bozo -- BOOM! (20s, x1)
-└── m1                                                          [OK] status nominal (2s, x0)`,
+		agent
+		├── m1
```
did we loose the sort order?

tommyp1ckles replied:
This might be related to how we query the db, let me take another look at this.

tommyp1ckles replied:
Changed up how we do the remote statedb query, which will result in the status' to be sent in alphabetical order (sorted by the identifier key).  

~Do you think the formatting code also ensure that the output is sorted?  Currently the tree will be built relative to the order in which the the Status slice is passed.~

@derailed Fixed this to sort by identifier prior to building the tree, this may be a bit redundant but it'll keep tests and output consistent.

joamaki commented on pkg/healthv2/provider.go:
```diff
@@ -0,0 +1,255 @@
+type HealthProvider struct {
+	db           *statedb.DB
+	stopped      atomic.Bool
+	StatusTable  statedb.RWTable[types.Status]
```
I would not export `RWTable` but keep it private to the provider. 

I would also make the `Table[types.Status]` available separately from `*HealthProvider`, and also have the indexes as top-level variables instead embedded into the health provider. The idea being that we would try to not "hide" the read-only access to statedb tables as they're central to how components using them work. E.g. it's easier to understand the health stuff when you see the table and `types.Status`.

joamaki commented on pkg/healthv2/provider.go:
```diff
@@ -0,0 +1,255 @@
+	LevelIndex   statedb.Index[types.Status, types.Level]
+}
+
+const HealthTable = "health"
```
For consistency with definitions elsewhere we should call this `HealthTableName`.

joamaki commented on pkg/healthv2/cell.go:
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package healthv2
```
I think we could move this into `pkg/hive/health` once the relocation of `pkg/hive` to `cilium/hive` is done. We'll anyway have `pkg/hive` that has the `cilium/cilium` specific wrapper for `hive.New` etc. That wrapper would then by default include stuff like health, statedb and so on.

derailed commented on pkg/healthv2/types/types.go:
```diff
@@ -0,0 +1,87 @@
+	return strings.Join(i, ".")
+}
+
+type Provider interface {
```
nit missing docs on most exports in this file

tommyp1ckles replied:
Add docs for relevant exported vars (skipping things like implementation of fmt.Stringer etc).

derailed commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -547,18 +549,49 @@ func (h testHelper) expectRetried(id uint64) {
-		bs, _ := status.JSON()
-		os.Stdout.Write(bs)
-		require.Failf(h.t, "health mismatch", "expected health level %q, got: %q (%s)", level, status.Level(), status.String())
+		//status, _ := h.health.Get([]string{"test"})
```
nit: axe?

derailed commented on pkg/statedb/reconciler/reconciler_test.go:
```diff
@@ -547,18 +549,49 @@ func (h testHelper) expectRetried(id uint64) {
-		require.Failf(h.t, "health mismatch", "expected health level %q, got: %q (%s)", level, status.Level(), status.String())
+		//status, _ := h.health.Get([]string{"test"})
+		tx := h.db.ReadTxn()
+		//iter, _ := h.health.StatusTable.All(tx)
```
nit: axe?

derailed reviewed (commented):
@tommyp1ckles Thanks for the updates!

learnitall reviewed (approved):
Looks good on docs, thanks! Up to you, but I think it would be a good idea to add a quick note in the upgrade notes about the breaking changes in `cilium-dbg` usage.

pippolo84 reviewed (approved):
Operator changes LGTM, thanks :100:

qmonnet reviewed (changes requested):
> Looks good on docs, thanks! Up to you, but I think it would be a good idea to add a quick note in the upgrade notes about the breaking changes in `cilium-dbg` usage.

+1, please do add a note. Doc changes look good otherwise.

tommyp1ckles committed:
Documentation: add upgrade notes for cilium-dbg status.

Do to the change of health provider backend, there are some
changes to how data is displayed that may be relevant to users
upgrading to v1.16.

As well, health status data is now strictly sorted by the fully
qualified health status identifier (i.e. [fq-module-id].[fq-component-id]).

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

derailed reviewed (approved):
@tommyp1ckles Thanks for the updates!

qmonnet reviewed (approved):
Thanks!
